QIR specification update 1 (#67)
microsoft/qsharp-language: Specifications/QIR/Data-Types.md

## Data Type Representation

We define LLVM representations for a variety of classical and quantum data types.

> QIR does not expect the runtime to provide garbage collection. Instead, it specifies a set of runtime functions that can be used by the language specific compiler to implement a reference counting scheme if needed, if the source language requires automatic memory management. See the section of [reference and alias counting](#reference-and-alias-counting) for more detail.

Reviewer: I might say "require" instead of "expect". I might also add a sentence saying that a runtime implementation that provides fully automated garbage collection is free to ignore (or strip out) the reference counting calls.

Author: Changed to require, and the sentence you suggested is already there further below in the section.

We define LLVM representations for a variety of classical and quantum data types.

> Representing the types used for qubits and measurement results as pointers to opaque LLVM structure types allows each target to provide a structure definition appropriate for that target.

Reviewer: ```suggestion
Representing the types used for qubits and measurement results as pointers to
opaque LLVM structure types allows each target to provide a structure definition
appropriate for that target.
```

A minor point: keeping the line length limited makes it easier to read diffs (PRs).

### Reference and Access Counting

> QIR specifies a set of runtime functions for types that are represented as pointers that may be used by the language specific compiler to expose them as immutable types in the language. 

Reviewer: ```suggestion
QIR specifies a set of runtime functions for types that are represented as pointers that may be used by the language-specific compiler to expose them as immutable types in the language. 
```

> - Runtime routines that create a new instance always initialize the instance

Reviewer 2: How the user count should be initialized?

Author: I'll add a sentence that the user count is supposed to be initialized to zero at the corresponding create runtime function.

> 

Reviewer 3: Can you please also address the following?

1. I feel the discrepancy for the PaulyY and PaulyZ values between the fragment 
> 0 is PauliI, 1 is PauliX, _**3**_ is PauliY, and _**2**_ is PauliZ

and the fragment (sorry, the `@` causes some unexpected formatting)
> @PauliI = constant i2 0
@PauliX = constant i2 1
@PauliY = constant i2 _**-1**_
@PauliZ = constant i2 _**-2**_
2. I propose the following change.
```diff
2c2
< 0..2..4 = {0. 2. 4}
---
> 0..2..4 = {0, 2, 4}
```

Reviewer 3: I understood the item 1. I propose the following clarification in [this file](https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Data-Types.md):
```diff
@@ -86,8 +86,8 @@ The following global constants are defined for use with the `%Result` and `%Paul
 
 @PauliI = constant i2 0
 @PauliX = constant i2 1
-@PauliY = constant i2 -1
-@PauliZ = constant i2 -2
+@PauliY = constant i2 -1    ; The value 3 (binary 11) is displayed as a 2-bit sigend value of -1 (binary 11).
+@PauliZ = constant i2 -2    ; The value 2 (binary 10) is displayed as a 2-bit sigend value of -2 (binary 10).
```

Author: Thanks for catching the dot, and I added the comment.

| Function                          | Signature                | Description |
|-----------------------------------|--------------------------|-------------|
| __quantum__rt__result_equal       | `i1(%Result*, %Result*)` | Returns true if the two results are the same, and false if they are different. |

> | __quantum__rt__result_update_reference_count   | `void(%Result*, i64)` | Adds the given integer value to the reference count for the result. Deallocates the result if the reference count becomes 0. The behavior is undefined if the reference count becomes negative. |

Reviewer 2: I've recently standardized the runtime to use i32 for ref/alias counts. Any reason we should be supporting i64?

Author: No, no reason besides that we are using i64 by default throughout QIR.

Author: I think you are right, i32 is enough and we are already using it in other places. I updated to i32.

### Tuples and User-Defined Types

Tuple data, including values of user-defined types, is represented as the corresponding LLVM structure type.
For instance, a tuple containing two integers, `(Int, Int)`, would be represented in LLVM as `type {i64, i64}`.

When [invoking callable values](https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Callables.md#invoking-a-callable-value) using the `__quantum__rt__callable_invoke` runtime function, 
tuples are passed as a pointer to an opaque LLVM structure, `%Tuple`. The pointer is expected to point to the contained data such that it can be cast to the correct data structures by the
receiving code.

> This permits to define runtime functions that are common for all tuples such as the ones listed below.

Reviewer: ```suggestion
This allows the definition of runtime functions that are common for all tuples such as the ones listed below.
```

> This permits to define runtime functions that are common for all tuples such as the ones listed below.

Reviewer: ```suggestion
This permits the definition of runtime functions that are common for all tuples, such as the functions listed below.
```

| __quantum__rt__tuple_copy      | `%Tuple*(%Tuple*, i1)`  | Creates a shallow copy of the tuple if the alias count is larger than 0 or the second argument is `true`. Returns the given tuple pointer otherwise, after increasing its reference count by 1. The reference count of the tuple items remains unchanged. |
| __quantum__rt__tuple_update_reference_count   | `void(%Tuple*, i64)` | Adds the given integer value to the reference count for the tuple. Deallocates the tuple if the reference count becomes 0. The behavior is undefined if the reference count becomes negative. |
| __quantum__rt__tuple_update_alias_count | `void(%Tuple*, i64)` | Adds the given integer value to the alias count for the tuple. Fails if the count becomes negative. |

### Unit

For source languages that include a unit type, the representation of this type
in LLVM depends on its usage.
If used as a return type for a callable, it should be translated into an LLVM
`void` function.
If it is used as a value, for instance as an element of a tuple, it should be represented as a null tuple pointer.

### Arrays

Within QIR, arrays are represented and passed around as a pointer to an opaque LLVM structure, `%Array`. 

> How to represent array data, i.e. what that pointer points to, is at the discretion of the runtime. All array manipulations, including item access, hence need to be performed by invoking the corresponding runtime function(s).

Reviewer: ```suggestion
How array data is represented, i.e., what that pointer points to, is at the discretion of the runtime. All array manipulations, including item access, hence need to be performed by invoking the corresponding runtime function(s).
```

receiving code.
This permits to define runtime functions that are common for all tuples.
For instance, this convention is used for callable wrapper functions; see
[below](#callable-values-and-wrapper-functions).

Many languages provide immutable tuples, along with operators that allow a modified copy of an existing tuple to be created.
QIR supports this by requiring the runtime to track and be able to access the following given a `%Tuple*`:
- The size of the tuple in bytes
- The alias count indicating how many handles to the tuple exist in the source code

The language specific compiler is responsible for injecting calls to increase and decrease the alias count as needed, as well as to accurately reflect when references to the LLVM structure representing a tuple are created and removed. 
See the section [above](#reference-and-alias-counting) regarding the distinction between alias and reference counting. 

In the case where the source language treats tuples as immutable values, the language-specific compiler is expected to request the necessary copies prior to modifying the tuple in place. 

> This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the alias count is 0 and it is hence save to modify the tuple in place.

Author: thanks

### Tuples and User-Defined Types

Tuple data, including values of user-defined types, is represented as the corresponding LLVM structure type.
For instance, a tuple containing two integers, `(Int, Int)`, would be represented in LLVM as `type {i64, i64}`.

> When passed to a callable function, tuples are passed as a pointer to an opaque LLVM structure, `%Tuple`. The pointer is expected to point to the contained data such that it can be cast to the correct data structures by the

Reviewer: In retrospect, this wording ("callable function") is ambiguous. I think it would be better to say "when passed to the callable_invoke runtime function..." or something like that, to be painfully precise.

Author: rephrased

receiving code.
This permits to define runtime functions that are common for all tuples.
For instance, this convention is used for callable wrapper functions; see
[below](#callable-values-and-wrapper-functions).

Many languages provide immutable tuples, along with operators that allow a modified copy of an existing tuple to be created.

> QIR permits to efficiently support this by requiring the runtime to track and be able to access the following given a `%Tuple*`:

Reviewer: ```suggestion
QIR supports this by requiring the runtime to track and be able to access the following given a `%Tuple*`:
```

> - The size of the tuple in number of bytes

Reviewer: ```suggestion
- The size of the tuple in of bytes
```

> - The user count indicating how many handles to the tuple exist in the source code

Reviewer: I'm thinking "alias" might be a better word than "user".

Reviewer 2: I like the idea! It also makes sense for the default "0" value of the count: there is a user, but no aliases.

Author: changed to alias

The language specific compiler is responsible for injecting calls to increase and decrease the user count as needed, as well as to accurately reflect when references to the LLVM structure representing a tuple are created and removed. 
See the section [above](#reference-and-access-counting) regarding the distinction between access and reference counting. 

> In the case where the source language exposes tuples as value types rather than reference types, the language specific compiler is expected to request the necessary copies prior to modifying the tuple in place. 

Reviewer: ```suggestion
In the case where the source language treats tuples as immutable values, the language-specific compiler is expected to request the necessary copies prior to modifying the tuple in place. 
```

> This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the user count is 0 and it is hence save to modify the tuple in place.

Reviewer: ```suggestion
This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the user count is 0 and it is therefore safe to modify the tuple in place.
```

Also, why do we need a "forced" copy? For that case, why wouldn't the compiler just generate the proper sequence of loads and stores (and user/ref count updates for inner pointers)?

Reviewer 2: There will be legitimate cases when a client might want to copy even in absence of aliases, no?

Author: Yes, a compiler may choose not to track alias counts and still need the option to have a copy be made.

This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the user count is 0 and it is hence save to modify the tuple in place.

The following utility functions are provided by the classical runtime to support tuples and user-defined types:

| Function                         | Signature             | Description |
|----------------------------------|-----------------------|-------------|
| __quantum__rt__tuple_create      | `%Tuple*(i64)`  | Allocates space for a tuple requiring the given number of bytes, sets the reference count to 1 and the user count to 0. |
| __quantum__rt__tuple_add_user   | `void(%Tuple*)` | Increases the current user count by one. |
| __quantum__rt__tuple_remove_user | `void(%Tuple*)` | Decreases the current user count by one, fails if the user count becomes negative. |

> | __quantum__rt__tuple_copy      | `%Tuple*(%Tuple*, i1)`  | Creates a shallow copy of the tuple if the user count is larger than 0 or the second argument is `true`. |

Reviewer 2: Perhaps, could explicitly say here that the calling code is responsible for proper reference counts updates on the contents of the tuple, because the runtime cannot do that?

Author: I added a sentence that the reference count of tuple items remains unchanged, and also added a similar sentence to the other copy functions.

| __quantum__rt__array_get_element_ptr_2d | `i8*(%Array*, i64, i64)`      | Returns a pointer to the element of the array at the zero-based indices given by the two `i64` arguments. |
| __quantum__rt__array_get_element_ptr | `i8*(%Array*, i64*)`             | Returns a pointer to the indicated element of the array. The `i64*` should point to an array of `i64`s that are the indices for each dimension. |
| __quantum__rt__array_slice       | `%Array*(%Array*, i32, %Range)`      | Creates and returns an array that is a slice of an existing array. The `i32` indicates which dimension the slice is on, which must be smaller than the number of dimensions in the array. The `%Range` specifies the slice. |
| __quantum__rt__array_project     | `%Array*(%Array*, i32, i64)`         | Creates and returns an array that is a projection of an existing array. The `i32` indicates which dimension the projection is on, and the `i64` specifies the specific index value to project. |

There are special runtime functions defined for allocating or releasing an
array of qubits.
See [here](Quantum-Runtime.md#qubit-management-functions) for these functions.

### Reference and Alias Counting

> QIR specifies a set of runtime functions for types that are represented as pointers that may be used by the language-specific compiler to expose them as immutable types in the language. 

Reviewer: A minor nit is that these functions don't apply to qubits, which are represented by pointers. It might be worth pointing out that this section doesn't apply to qubits, just for clarity.

Author: I added a sentence.

To ensure that unnecessary copying of data can be avoided, QIR distinguishes two kinds of counts that can be tracked: reference counts and alias counts. 

> Reference counts track the number of handles that allow to access a certain value *in LLVM*. They hence determine when the value can be released by the runtime; values are allocated with a reference count of 1, and will be released when their reference count reaches 0. 

Reviewer: ```suggestion
Reference counts track the number of handles that allow access to a certain value *in LLVM*. They hence determine when the value can be released by the runtime; values are allocated with a reference count of 1, and will be released when their reference count reaches 0. 
```

> Alias counts, on the other hand, track how may handles exist *in the source language*. 

Reviewer: ```suggestion
Alias counts, on the other hand, track how many handles to a value exist *in the source language*. 
```

> They determine when the runtime needs to copy data; when copy functions are invoked, the copy is executed only if the alias count is larger than 0, or the copy is explicitly forced. Alias counts are hence useful to optimize the handling of data types that are represented as pointers in QIR, but are value types, i.e. immutable, within the source language. 

Reviewer: ```suggestion
They determine when the runtime needs to copy data; when copy functions are invoked, the copy is executed only if the alias count is larger than 0, or the copy is explicitly forced. Alias counts are useful for optimizing the handling of data types that are represented as pointers in QIR, but are value types, i.e. immutable, within the source language. 
```

> The compiler is responsible to track both counts as needed by injecting the corresponding calls to modify them. A call to modify such counts will only ever modify the count for the given instance itself and not for inner item such as e.g. tuple or array items, or captured values for callables; the compiler is responsible for injecting calls to update counts for inner items as needed.

Reviewer: ```suggestion
The compiler is responsible for generating code that tracks both counts correctly by injecting the corresponding calls to modify them. A call to modify such counts will only ever modify the count for the given instance itself and not for any inner items such as an element of a tuple or an array or a value captured by a callable; the compiler is responsible for injecting calls to update counts for inner items as needed.
```

> A targeted runtime is free to provide other mechanism for garbage collection and treat calls to modify reference counts as hints or as simple no-ops.

Reviewer: ```suggestion
A runtime implementation is free to provide another mechanism for garbage collection and to treat calls to modify reference counts as hints or as simple no-ops.
```

- Runtime routines that create a new instance always initialize the instance
  with a reference count of 1, and an alias count of 0.
- For each pointer type, with the exception of `%Qubit*`, 

>   a runtime function ending in `_update_reference_count` exists that can be used to modify the reference count of an instance as needed. If the reference count reaches 0, the instance may be released. Decreasing the reference count below 0 or accessing a value after its reference count has reached 0 results in undefined behavior.

Reviewer: A philosophical note: I really dislike the notion of "undefined behavior" since it tends to result in the growth of unplanned spec items: a couple of influential implementations happen to pick the same behavior for some case that's undefined, and all of a sudden people are writing code that relies on that behavior.
I would much prefer to fail when something nonsensical occurs (negative reference count).
In the case of accessing after release, there's no way to force any particular behavior, so "undefined" is the best we can do.

Author: I agree with that I would rather require a proper failure, but I think that is too much to ask, given that the runtime should be allowed to free and forget that memory.

- For all data types that support a runtime function to create a shallow copy, 

>   a runtime function ending in `_update_alias_count` exists that can be used to modify the alias count of an instance as needed. These functions exist for `%Tuple*`, `%Array*`, and `%Callable*` types. The alias count can never be negative; decreasing the alias count below 0 results in a runtime exception.

Reviewer: ```suggestion
  a runtime function ending in `_update_alias_count` exists that can be used to modify the alias count of an instance as needed. These functions exist for `%Tuple*`, `%Array*`, and `%Callable*` types. The alias count can never be negative; decreasing the alias count below 0 results in program failure.
```
We don't have exceptions.

Author: Well, but the runtime has exceptions (usually), so I was thinking of requiring the runtime to throw. I am not sure what handling we would want to require if we didn't want to fail execution?

Reviewer: I would require failure. We require failure in some other situations, I believe.

> - The function to modify reference and alias count accept a 

Reviewer: ```suggestion
- The functions that modify reference and alias count should accept a 
```

>   null instance pointer and should simply ignore the call if the pointer is null.

Reviewer: ```suggestion
  null instance pointer and simply ignore the call if the pointer is null.
```
